File: tests/browser/unit/arranque/test_a_first_run_does_not_wear_the_previous_install.py
"""A first run does not wear the previous install's browser state (V2-735).

The operator, on a freshly reset install: «cuando el sistema arranca, no quiero
que por defecto el orbe esté metido en la barra inferior. Quiero que el orbe esté
desplegado y que se vea el orbe grande con el pulso y todo… totalmente arrancado,
porque la voz tiene que empezar a sonar enseguida».

And by default it IS: `store.orbDock` defaults to "eye" and `store.powerOff` to
false. What he was looking at was the PREVIOUS install's `hb_orb_dock=bar` and
`hb_power_off=1` — a factory reset (V2-670) starts the AGENT over and cannot
reach localStorage, which is per-origin and survives everything the server can do
to itself.
"""
from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

import pytest

SCRIPT = Path(__file__).with_name("test_a_first_run_does_not_wear_the_previous_install.mjs")
MAIN = Path("frontend/app/main.js")
STORE = Path("frontend/app/core/store.js")


@pytest.mark.skipif(shutil.which("node") is None, reason="node not installed")
def test_the_takeover_wipes_what_is_ours_and_nothing_else():
    r = subprocess.run(["node", str(SCRIPT)], capture_output=True, text=True, timeout=60)
    assert r.returncode == 0, (r.stdout or "") + (r.stderr or "")


def test_the_deployed_orb_and_a_live_voice_ARE_the_defaults():
    """Nothing in this fix changes what the product defaults to — the point is that a default is what a
    new operator should get. If either of these ever flips, the fix above is papering over it."""
    src = STORE.read_text(encoding="utf-8")
    assert 'localStorage.getItem("hb_orb_dock") === "bar" ? "bar" : "eye"' in src, (
        "the orb lives in the big eye cluster unless the operator moved it")
    assert 'localStorage.getItem("hb_power_off") === "1"' in src, (
        "the voice is ON unless the operator switched it off")


def test_the_takeover_runs_on_the_first_run_branch_and_stops_the_boot():
    """It has to sit exactly where the engine says «nobody has ever chosen a language», and the caller
    has to RETURN when it fires: the page is on its way out, and opening the language modal on a document
    that is being replaced is how a first run shows a flash of the wrong thing."""
    src = MAIN.read_text(encoding="utf-8")
    i_chosen = src.index('s.chosen !== false')
    i_take = src.index("firstRun.takeoverOnFirstRun(", i_chosen)
    i_open = src.index("store.setLangOnboardOpen(true)", i_chosen)
    assert i_chosen < i_take < i_open, "wipe first, then the picker"
    between = src[i_take:i_open]
    assert "return" in between, "a takeover must stop the boot: the page is going away"

File: tests/browser/unit/arranque/test_a_first_run_does_not_wear_the_previous_install.mjs
// ============================================================================
// test_a_first_run_does_not_wear_the_previous_install.mjs — V2-735, node 4.195.
//
// THE OPERATOR'S REPORT (2026-09-20, on a freshly reset install): «cuando el
// sistema arranca, no quiero que por defecto el orbe esté metido en la barra
// inferior. Quiero que el orbe esté desplegado y que se vea el orbe grande con
// el pulso y todo… totalmente arrancado, porque la voz tiene que empezar a
// sonar enseguida.»
//
// And by default it IS: `store.orbDock` defaults to "eye", `store.powerOff` to
// false. What he was looking at was the PREVIOUS install's `hb_orb_dock=bar`
// and `hb_power_off=1`, still in the browser — a factory reset starts the agent
// over and cannot reach localStorage.
//
// Run: node tests/browser/unit/arranque/test_a_first_run_does_not_wear_the_previous_install.mjs
// ============================================================================
import assert from "node:assert/strict";

function fakeStorage(seed = {}) {
  const m = new Map(Object.entries(seed));
  return {
    get length() { return m.size; },
    key: (i) => [...m.keys()][i] ?? null,
    getItem: (k) => (m.has(k) ? m.get(k) : null),
    setItem: (k, v) => m.set(k, String(v)),
    removeItem: (k) => m.delete(k),
    _keys: () => [...m.keys()],
  };
}

const { takeoverOnFirstRun, clearInheritedViewState, isOurs } =
  await import("../../../../frontend/app/core/first-run.js?v=1");

// ── 1. the operator's own case: the shape of the install before this one ────────────────────────────────
{
  const local = fakeStorage({
    hb_orb_dock: "bar",            // the orb docked in the bottom bar
    hb_power_off: "1",             // and the voice off, so nothing ever speaks
    hb_desktop: '{"cards":[1,2]}', // somebody else's cards
    hb_lang: "de",
    zaelar_mic: "some-device",
    unrelated_key: "keep me",      // not ours: another app on the same origin
  });
  const session = fakeStorage();
  let reloaded = 0;
  const took = takeoverOnFirstRun({ local, session, reload: () => reloaded++ });

  assert.equal(took, true, "THE BUG: a first run came up wearing the previous install's state");
  assert.equal(reloaded, 1, "the signals are seeded at import — only a reload can undo them");
  assert.deepEqual(local._keys(), ["unrelated_key"],
    `our namespace must be gone and nothing else touched — ${local._keys()}`);
}

// ── 2. a genuinely new browser: nothing to clear, nothing to reload ─────────────────────────────────────
{
  const local = fakeStorage({ unrelated_key: "x" });
  const session = fakeStorage();
  let reloaded = 0;
  assert.equal(takeoverOnFirstRun({ local, session, reload: () => reloaded++ }), false);
  assert.equal(reloaded, 0, "a clean install must not blink through an extra page load");
  assert.deepEqual(local._keys(), ["unrelated_key"]);
}

// ── 3. never twice in one tab, whatever gets rewritten in between ───────────────────────────────────────
{
  const local = fakeStorage({ hb_orb_dock: "bar" });
  const session = fakeStorage();
  let reloaded = 0;
  const reload = () => { reloaded++; local.setItem("hb_lang", "es"); };   // boot rewrites a key of ours
  assert.equal(takeoverOnFirstRun({ local, session, reload }), true);
  assert.equal(takeoverOnFirstRun({ local, session, reload }), false,
    "a second attempt in the same tab is how this becomes a reload loop");
  assert.equal(reloaded, 1);
}

// ── 4. no sessionStorage → no guard → do not risk a loop ────────────────────────────────────────────────
{
  const local = fakeStorage({ hb_orb_dock: "bar" });
  const blind = { getItem() { throw new Error("blocked"); }, setItem() { throw new Error("blocked"); } };
  let reloaded = 0;
  assert.equal(takeoverOnFirstRun({ local, session: blind, reload: () => reloaded++ }), false);
  assert.equal(reloaded, 0, "without a guard a reload could repeat forever — do nothing instead");
}

// ── 5. what counts as ours ──────────────────────────────────────────────────────────────────────────────
for (const k of ["hb_theme", "hb_orb_dock", "zaelar_mic"]) assert.ok(isOurs(k), k);
for (const k of ["theme", "lang", "", null, "myapp_hb_theme"]) assert.ok(!isOurs(k), String(k));

// ── 6. a storage that throws mid-sweep leaves the boot alone ────────────────────────────────────────────
{
  const angry = { get length() { throw new Error("private window"); } };
  assert.deepEqual(clearInheritedViewState(angry), [], "a blocked store has nothing to inherit either");
}

// ── 7. ANY reset, not only a first run (operator, 2026-09-25) ───────────────────────────────────────────
// «en cuanto hago un reset, todas esas variables de estado pasan al estado inicial». A reset that keeps the
// language never reaches takeoverOnFirstRun, so the mic and speaker came back muted. The reset's EPOCH does.
{
  const { takeoverOnReset } = await import("../../../../frontend/app/core/first-run.js?v=1");
  const local = fakeStorage({ hb_wipe: "100", hb_mic_muted: "1", hb_bot_muted: "1", hb_orb_dock: "bar",
                              unrelated_key: "keep me" });
  const session = fakeStorage();
  let reloaded = 0;
  const took = await takeoverOnReset({ fetchEpoch: async () => 200, local, session, reload: () => reloaded++ });
  assert.equal(took, true, "THE BUG: a reset left the previous mute switches in the browser");
  assert.equal(reloaded, 1);
  assert.deepEqual(local._keys().sort(), ["hb_wipe", "unrelated_key"], local._keys());
  assert.equal(local.getItem("hb_wipe"), "200", "the epoch obeyed is remembered, so it happens once");

  // the same epoch again (next page load): nothing
  local.setItem("hb_mic_muted", "1");            // he mutes on purpose AFTER the reset
  assert.equal(await takeoverOnReset({ fetchEpoch: async () => 200, local, session: fakeStorage(),
                                       reload: () => reloaded++ }), false);
  assert.equal(local.getItem("hb_mic_muted"), "1", "a choice made after the reset is his, and it stays");
  assert.equal(reloaded, 1);
}
{
  const { takeoverOnReset } = await import("../../../../frontend/app/core/first-run.js?v=1");
  // a brand-new browser: records the epoch, wipes nothing, never blinks
  const local = fakeStorage({ unrelated_key: "x" });
  let reloaded = 0;
  assert.equal(await takeoverOnReset({ fetchEpoch: async () => 300, local, session: fakeStorage(),
                                       reload: () => reloaded++ }), false);
  assert.equal(reloaded, 0);
  assert.equal(local.getItem("hb_wipe"), "300");
  // no epoch served, or the engine unreachable: nothing happens
  const l2 = fakeStorage({ hb_mic_muted: "1" });
  assert.equal(await takeoverOnReset({ fetchEpoch: async () => 0, local: l2, session: fakeStorage(), reload() {} }), false);
  assert.equal(await takeoverOnReset({ fetchEpoch: async () => { throw new Error("down"); }, local: l2,
                                       session: fakeStorage(), reload() {} }), false);
  assert.equal(l2.getItem("hb_mic_muted"), "1");
}

console.log("ok: a first run does not wear the previous install");
